src/Backends/PrefersColorSchemeSettings.vala

```diff
@@ -0,0 +1,144 @@
+        Object (accounts_service: accounts_service);
+    }
+
+    construct {
+        color_settings = new GLib.Settings ("io.elementary.settings-daemon.prefers-color-scheme");
+
+        color_settings.changed["prefer-dark-schedule"].connect (update);
+
+        update ();
+    }
+
+    private void update () {
+        var schedule = color_settings.get_string ("prefer-dark-schedule");
+
+        if (schedule == "sunset-to-sunrise") {
+            get_location.begin ((obj, res) => {
+                async_result = res;
+            });
+
+            start_timer ();
+        } else if (schedule == "manual") {
+            start_timer ();
+        } else {
+            get_location.end (async_result);
+            stop_timer ();
+        }
+    }
+
+    private void start_timer () {
+        if (!time_running) {
+            var time = new TimeoutSource (1000);
+            time.set_callback (time_callback);
+            time_id = time.attach (null);
+            time_running = true;
+        }
+    }
+
+    private void stop_timer () {
+        if (time_running) {
+            time_running = !Source.remove (time_id);
+        }
+    }
+
+    private async void get_location () {
+        try {
+            var simple = yield new GClue.Simple (Build.PROJECT_NAME, GClue.AccuracyLevel.CITY, null);
+
+            simple.notify["location"].connect (() => {
+                on_location_updated (simple.location.latitude, simple.location.longitude);
+            });
+
+            on_location_updated (simple.location.latitude, simple.location.longitude);
+        } catch (Error e) {
+            warning ("Failed to connect to GeoClue2 service: %s", e.message);
+            return;
+        }
+    }
+
+    private bool time_callback () {
+        var schedule = color_settings.get_string ("prefer-dark-schedule");
+
+        var now = new DateTime.now_local ();
+        double from, to;
+        if (schedule == "sunset-to-sunrise") {
+            from = 20.0;
+            to = 6.0;
+
+            SettingsDaemon.Utils.SunriseSunsetCalculator.get_sunrise_and_sunset (now, pos_lat, pos_long, out to, out from);
+        } else if (schedule == "manual") {
+            from = color_settings.get_double ("prefer-dark-schedule-from");
+            to = color_settings.get_double ("prefer-dark-schedule-to");
+        } else {
+            return true;
+        }
+
+        var is_in = is_in_time_window (date_time_double (now), from, to);
+        var new_color_scheme = Granite.Settings.ColorScheme.NO_PREFERENCE;
+        if (is_in) {
+            new_color_scheme = Granite.Settings.ColorScheme.DARK;
+        }
+
+        if (new_color_scheme == accounts_service.prefers_color_scheme) {
+            return true;
+        }
+
+        accounts_service.prefers_color_scheme = new_color_scheme;
+
+        return true;
+    }
+
+    private void on_location_updated (double latitude, double longitude) {
```

Comment:
I think we probably want to recalculate sunrise and sunset and update the `from` and `to` time when the location is updated?

Reply:
I now recalculate `sunrise` and `sunset` when the location is updated.  `from` and `to` are set based on `sunset` and `sunrise` in `time_callback ()`. I also added a cache for last coordinates and manually call it `on_location_updated` once.